bokehjs/src/lib/models/renderers/contour_renderer.ts

```diff
@@ -0,0 +1,88 @@
+import {DataRenderer, DataRendererView} from "./data_renderer"
+import {GlyphRenderer, GlyphRendererView} from "./glyph_renderer"
+import {Renderer} from "./renderer"
+import {GlyphView} from "../glyphs/glyph"
+import {MultiLine} from "../glyphs/multi_line"
+import {MultiPolygons} from "../glyphs/multi_polygons"
+import {ColumnDataSource} from "../sources/column_data_source"
+import * as p from "core/properties"
+import {build_view} from "core/build_views"
+import {SelectionManager} from "core/selection_manager"
+
+export class ContourRendererView extends DataRendererView {
+  override model: ContourRenderer
+
+  fill_view: GlyphRendererView
+  line_view: GlyphRendererView
+
+  get glyph_view(): GlyphView {
+    if (this.fill_view.glyph.data_size > 0)
+      return this.fill_view.glyph
+    else
+      return this.line_view.glyph
+  }
+
+  override async lazy_initialize(): Promise<void> {
+    await super.lazy_initialize()
+    const {parent} = this
+    const {fill_renderer, line_renderer} = this.model
+    this.fill_view = await build_view(fill_renderer, {parent})
+    this.line_view = await build_view(line_renderer, {parent})
+  }
+
+  override remove(): void {
+    this.fill_view.remove()
+    this.line_view.remove()
+    super.remove()
+  }
+
+  protected _render(): void {
+    this.fill_view.render()
+    this.line_view.render()
+  }
+
+  override renderer_view<T extends Renderer>(renderer: T): T["__view_type__"] | undefined {
+    if (renderer instanceof GlyphRenderer) {
+      if (renderer == this.fill_view.model)
+        return this.fill_view
+      if (renderer == this.line_view.model)
+        return this.line_view
+    }
+    return super.renderer_view(renderer)
+  }
+}
+
+export namespace ContourRenderer {
+  export type Attrs = p.AttrsOf<Props>
+
+  export type Props = DataRenderer.Props & {
+    fill_renderer: p.Property<GlyphRenderer>
+    line_renderer: p.Property<GlyphRenderer>
+    levels: p.Property<number[]>
+  }
+}
+
+export interface ContourRenderer extends ContourRenderer.Attrs {}
+
+export class ContourRenderer extends DataRenderer {
+  override properties: ContourRenderer.Props
+  override __view_type__: ContourRendererView
+
+  constructor(attrs?: Partial<ContourRenderer.Attrs>) {
+    super(attrs)
+  }
+
+  static {
+    this.prototype.default_view = ContourRendererView
+
+    this.define<ContourRenderer.Props>(({Array, Number, Ref}) => ({
+      fill_renderer: [ Ref(GlyphRenderer), () => new GlyphRenderer({glyph: new MultiPolygons(), data_source: new ColumnDataSource()}) ],
```

Comment:
I needed to explicitly include the defaults here otherwise I get an error about different defaults in python and javascript. But similar code is not needed in `GraphRenderer` despite it being very similar?

Reply:
I'm not sure offhand @mattpap might know

Reply:
See
https://github.com/bokeh/bokeh/blob/b85d0cd2a52d81d0a64404e0fcd6ff1446a236be/bokehjs/test/defaults/index.ts#L60-L62

But I would ask if you really need a default here, especially if there are higher level APIs that will be preferable to use by the user. If this is legitimate, then sure keep it, but don't force yourself to fill in defaults. Bokeh (3.0) will be perfectly happy with objects whose properties have unset values, until serialization time.

Reply:
Thanks for pointing that out. I'd prefer not to use a default so I will look into it further.